components/ingest-service/pipeline/processor/action_project_tagger_test.go

```diff
@@ -0,0 +1,416 @@
+	inbox <- message.NewChefAction(context.Background(), &chef.Action{}, errc)
+	close(inbox)
+
+	out := actionBundleProjectTagger(inbox, authzClient)
+
+	<-out
+	<-out
+	<-out
+	<-out
+	<-out
+
+	assert.Equal(t, 1, listProjectRulesCount)
+}
+
+// A simple run through of the bundle project tagger processor.
+// Two messages are sent through with only one matching a project.
+func TestActionBundlerMatchProjectRule(t *testing.T) {
+	inbox := make(chan message.ChefAction, 100)
+	testProjectName := "Test"
+	orgName := "org_1"
+	projectRules := map[string]*iam_v2.ProjectRules{}
+	projectRules[testProjectName] = &iam_v2.ProjectRules{
+		Rules: []*iam_v2.ProjectRule{
+			{
+				Type: iam_v2.ProjectRuleTypes_EVENT,
+				Conditions: []*iam_v2.Condition{
+					{
+						Type:   iam_v2.ProjectRuleConditionTypes_CHEF_ORGS,
+						Values: []string{orgName},
+					},
+				},
+			},
+		},
+	}
+	authzClient := iam_v2.NewMockProjectsClient(gomock.NewController(t))
+	authzClient.EXPECT().ListProjectRules(gomock.Any(), gomock.Any()).Return(
+		&iam_v2.ProjectCollectionRulesResp{ProjectRules: projectRules}, nil)
+	errc := make(chan error)
+
+	action1 := message.NewChefAction(context.Background(), &chef.Action{
+		OrganizationName: orgName,
+	}, errc)
+
+	action2 := message.NewChefAction(context.Background(), &chef.Action{
+		OrganizationName: "no_match",
+	}, errc)
+
+	inbox <- action1
+	inbox <- action2
+	close(inbox)
+
+	out := BuildActionProjectTagger(authzClient)(inbox)
+
+	processMsg1 := <-out
+	assert.Equal(t, []string{testProjectName}, processMsg1.InternalChefAction.Projects)
+
+	processMsg2 := <-out
+	assert.Equal(t, []string{}, processMsg2.InternalChefAction.Projects)
+}
+
+func TestActionBundlerMatchProjectRuleNodeRuleType(t *testing.T) {
+	inbox := make(chan message.ChefAction, 100)
+	testProjectName := "Test"
+	orgName := "org_1"
+	projectRules := map[string]*iam_v2.ProjectRules{}
+	projectRules[testProjectName] = &iam_v2.ProjectRules{
+		Rules: []*iam_v2.ProjectRule{
+			{
+				Type: iam_v2.ProjectRuleTypes_NODE,
+				Conditions: []*iam_v2.Condition{
+					{
+						Type:   iam_v2.ProjectRuleConditionTypes_CHEF_ORGS,
+						Values: []string{orgName},
+					},
+				},
+			},
+		},
+	}
+	authzClient := iam_v2.NewMockProjectsClient(gomock.NewController(t))
+	authzClient.EXPECT().ListProjectRules(gomock.Any(), gomock.Any()).Return(
+		&iam_v2.ProjectCollectionRulesResp{ProjectRules: projectRules}, nil)
+	errc := make(chan error)
+
+	action1 := message.NewChefAction(context.Background(), &chef.Action{
+		OrganizationName: orgName,
+	}, errc)
+
+	action2 := message.NewChefAction(context.Background(), &chef.Action{
+		OrganizationName: "no_match",
+	}, errc)
+
+	inbox <- action1
+	inbox <- action2
+	close(inbox)
+
+	out := BuildActionProjectTagger(authzClient)(inbox)
+
+	processMsg1 := <-out
+	assert.Equal(t, []string{}, processMsg1.InternalChefAction.Projects)
+
+	processMsg2 := <-out
+	assert.Equal(t, []string{}, processMsg2.InternalChefAction.Projects)
```

Comment:
I'm finding these tests a bit hard to understand, and I wonder if the tests wouldn't need to do as much if the logic for how to treat the messages and rules was separated from the channel handling /message passing code.

Reply:
🤔 I will look into this.

Reply:
I think this last test (TestActionBundlerMatchProjectRuleNodeRuleType) is not needed, so I removed it. This is being tested here https://github.com/chef/automate/pull/214/files/ca2de00b00ff7b701f275efcf0cbcfea39652b8c#diff-7e0662f4628cc26ca7731861f082208fR63.

Reply:
I went through and added more comments on what is going on in these tests. I do not know a way to break apart the channel handling from the bundler algorithm. This algorithm was designed to use the length of the channel to figure out how many messages to bundle together. Then the algorithm counts down each message as it goes through.